include/unifex/range_stream.hpp

```diff
@@ -59,26 +64,30 @@ struct next_sender {
 
   friend constexpr blocking_kind tag_invoke(
       tag_t<blocking>,
-      const stream&) noexcept {
+      const stream<Range>&) noexcept {
     return blocking_kind::always_inline;
   }
 
   template <typename Receiver>
-  operation<Receiver> connect(Receiver&& receiver) && {
-    return operation<Receiver>{stream_, (Receiver &&) receiver};
+  operation<Receiver, Range> connect(Receiver&& receiver) && {
+    return operation<Receiver, Range>{stream_, (Receiver &&) receiver};
   }
   template <typename Receiver>
   void connect(Receiver&& receiver) const& = delete;
 };
 
+template <std::ranges::range Range>
 struct stream {
-  int next_;
-  int max_;
+  using iterator_t = std::ranges::iterator_t<Range>;
+  using sentinel_t = std::ranges::sentinel_t<Range>;
+
+  iterator_t begin_;
+  sentinel_t end_;
 
-  explicit stream(int max) : next_(0), max_(max) {}
-  explicit stream(int start, int max) : next_(start), max_(max) {}
+  explicit stream(Range&& r) : begin_(std::begin(r)), end_(std::end(r)) {}
+  explicit stream(Range& r) : begin_(std::begin(r)), end_(std::end(r)) {}
```

Comment:
I'll need some help here. I thought I understood templated universal references, but clearly not. From my understanding some function
```
template <typename T>
void foo(T&& t) {}
```
Would be in a state where it would support both an lvalue reference and an rvalue reference. When that function is called `t` would either be type `T&` or `T&&`, but we do not know. So we have to use `std::forward<T>` to do a conditional rvalue cast. 

With this logic I assumed that the `stream(Range&&)` constructor would be enough to handle both lvalue and rvalue types, but for some reason it didn't and I needed ot make the `stream(Range&)` constructor.

Reply:
I can explain that. In this code:

```c++
template <typename T>
void foo(T&& t) {}
```

... the type `T` is deduced from the type of the argument. And you're correct, it could be an lvalue or an rvalue.

In this code, however:

```c++
template <typename T>
struct S {
  void foo(T&& t) {}
};
```

... the type `T` is _not_ deduced. It is specified when instantiating the class template as either an lvalue reference type or as an ordinary value type. If you specify it to be a value type -- say, `int` -- then the `T&&` in the argument list of `foo()` becomes `int&&`. That will bind _only_ to rvalue `int`s. It will _not_ bind to lvalues.

Hope this helps.

Reply:
Wow I really appreciate that explanation. So the gist is that if the template type parameters are deduced, then the _referencesness_ has different semantics than when they are not deduced. I'm going to have to make sure I retain this.